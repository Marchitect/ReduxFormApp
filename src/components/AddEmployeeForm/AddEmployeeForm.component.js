import React from "react"
import { useDispatch, useSelector } from "react-redux"
import { SAVE_FORMDATA } from "../../constants/action-types"
import Form from "react-bootstrap/Form"
import Col from "react-bootstrap/Col"
import Button from "react-bootstrap/Button"
import { Formik, Form as FormikForm } from "formik"
import * as Yup from "yup"
import { TextInput } from "../TextInput"

export function AddEmployeeForm() {
  const dispatch = useDispatch()
  // make sure the same salary number is not allowed to be added repeatedly
  const stateData = useSelector(state => state)
  let salaryNoArray = []
  return (
    <Formik
      initialValues={{
        salaryNo: "",
        employeeName: "",
        bsb: "",
        accountNo: ""
      }}
      validationSchema={Yup.object({
        salaryNo: Yup.string()
          // no space allowed at the beginning
          .matches(
            /^[a-zA-Z0-9-]+[a-zA-Z0-9 -]*$/,
            "Please enter a valid account number"
          )
          .required("This field is required"),
        employeeName: Yup.string()
          .matches(/^[a-zA-Z]+[a-zA-Z -]*$/, "Please enter a valid name")
          .max(50, "Name must be 50 characters or less")
          .required("This field is required"),
        bsb: Yup.string()
          .matches(
            /^(?!00)[0-9]{6}$/,
            "Please enter a six-digit number only, BSB cannot start with 00"
          )
          .required("This field is required"),
        accountNo: Yup.string()
          .matches(
            // \S means space, ^\S means no space at the beginning, \S near the end means no spaces anywhere
            /^\S(?=.*\d)(?=.*[1-9]).\S{6,10}$/,
            "Please enter a valid account number"
          )
          .required("This field is required")
      })}
      // variable data (or call it values because they are values from inputs) here is linked to the reducer, formData
      onSubmit={(data, { setSubmitting, resetForm }) => {
        stateData.formData.map((entry) => (
          salaryNoArray.push(entry.salaryNo)
        ))
        if (!salaryNoArray.includes(data.salaryNo)){
          // the variable "data" here needs to be kept with the same name in reducers
          dispatch({ type: SAVE_FORMDATA, data })   
          setTimeout(() => {
            alert(
              "Data is successfully submitted. " + JSON.stringify(data, null, 2)
            )
            setSubmitting(false)
          }, 400)
        }
        else {
          alert("Salary number cannot be repetitive.")
        }
        resetForm()
      }}
    >
      <FormikForm>
        <h3>Employee Details</h3>
        <Form.Row>
          <TextInput
            label="Reference / Salary number"
            name="salaryNo"
            type="text"
            placeholder="Reference or salary number"
            tooltip="true"
          />
        </Form.Row>
        <Form.Row>
          <TextInput
            label="Employee name"
            name="employeeName"
            type="text"
            placeholder="Employee name"
          />
        </Form.Row>
        <Form.Row>
          <TextInput label="BSB" name="bsb" type="text" placeholder="BSB" />
          <TextInput
            label="Account number"
            name="accountNo"
            type="text"
            placeholder="Account number"
          />
        </Form.Row>
        <Form.Row>
          <Col>
            <Button type="submit">Submit</Button>
          </Col>
        </Form.Row>
      </FormikForm>
    </Formik>
  )
}
