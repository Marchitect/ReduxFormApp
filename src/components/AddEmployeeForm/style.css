.error {
  color: red;
}
.text-input-error {
  border-color: red !important;
}
.form-row {
  margin-top: 15px;
}