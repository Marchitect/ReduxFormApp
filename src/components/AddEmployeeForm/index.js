// index.js
import { AddEmployeeForm } from './AddEmployeeForm.component'

export { AddEmployeeForm }