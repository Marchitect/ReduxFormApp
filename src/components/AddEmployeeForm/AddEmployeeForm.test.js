// AddEmployeeForm.test.js
import React from "react"
import Enzyme, { shallow } from "enzyme"
import Adapter from "enzyme-adapter-react-16"
import { AddEmployeeForm } from "./AddEmployeeForm.component"
import { render, fireEvent, wait } from '@testing-library/react';

Enzyme.configure({ adapter: new Adapter() })

jest.mock("react-redux", () => ({
  useDispatch: () => {}
}))

describe("AddEmployeeForm", () => {
  it("renders correctly", () => {
    const addEmployeeForm = shallow(<AddEmployeeForm />)
    expect(addEmployeeForm).not.toBeNull()
    expect(addEmployeeForm).toMatchSnapshot()
  })

  it("submits correct values", async () => {
    const mock = jest.fn()
    const { container } = render(<AddEmployeeForm onSubmit={mock} />)
    const salaryNo = container.querySelector('input[name="salaryNo"]')
    const employeeName = container.querySelector('input[name="employeeName"]')
    const bsb = container.querySelector('select[name="bsb"]')
    const accountNo = container.querySelector('select[name="accountNo"]')
    const submit = container.querySelector('button[type="submit"]')

    await wait(() => {
      fireEvent.change(salaryNo, {
        target: {
          value: "mockno123"
        }
      })
    })

    await wait(() => {
      fireEvent.change(employeeName, {
        target: {
          value: "Test Name"
        }
      })
    })

    await wait(() => {
      fireEvent.change(bsb, {
        target: {
          value: "011234"
        }
      })
    })

    await wait(() => {
      fireEvent.change(accountNo, {
        target: {
          value: "12345678"
        }
      })
    })

    await wait(() => {
      fireEvent.click(submit)
    })

    wait(() => {
      expect(mock).toBeCalled()
    })
  })
})
