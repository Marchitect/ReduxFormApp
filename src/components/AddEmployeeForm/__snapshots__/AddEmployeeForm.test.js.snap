// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddEmployeeForm renders correctly 1`] = `ShallowWrapper {}`;
